src/fsm.cpp

```diff
@@ -171,3 +179,108 @@ void fsm::FSM::validate_final_states() const {
     }
 }
 
+bool fsm::FSM::evaluate(const char* input) {
+    fsm::String word(input);
+
+    for(int i = 0, l = word.size(); i < l; i++){
+        FSM::transition(word[i] - '0'); // convert char to int
+    }
+
+    bool flag = FSM::is_in_final_state();
+    FSM::restart();
+
+    return flag;
+}
+
+fsm::FSM fsm::FSM::operator!() const {
+    std::vector<fsm::State> newFinalStates;
+    fsm::FSM complementMachine = *this;
+
+    for(int i = 0, sz = get_states_count(); i < sz; i++){
+        if(std::find(final_states_.begin(),final_states_.end(),states_[i]) == final_states_.end()){
+            newFinalStates.push_back(states_[i]);
+        }
+    }
+
+    complementMachine.set_final_states(newFinalStates);
+
+    return complementMachine;
+}
+
+fsm::FSM fsm::FSM::operator&(const fsm::FSM &rhs) const {
+    fsm::FSM intersectionMachine;  // Initiate empty machine.
+    fsm::State comboState = get_current_state() + rhs.get_current_state();  // Get the initial state.
+    for(int i = 0, sz = get_alphabet_count(); i < sz; i++) { intersectionMachine.add_symbol(i); }  // Add the alphabet.
+    intersectionMachine.add_state(comboState);
+
+    fsm::fill(*this, rhs, intersectionMachine, comboState);
+
+    intersectionMachine.set_initial_state(comboState);
+    intersectionMachine.restart();
+
+    return intersectionMachine;
+}
+
+void fsm::fill(fsm::FSM m1, fsm::FSM m2, fsm::FSM &m3, fsm::State prevState) {
+    for(int i = 0, sz = m1.get_alphabet_count(); i < sz; i++){
+        auto m1old = m1, m2old = m2;
+        auto m3States = m3.get_states();
+
+        m1.transition(i);
+        m2.transition(i);
+
+        fsm::State comboState = m1.get_current_state() + m2.get_current_state();
+
+        if(std::find(m3States.begin(), m3States.end(), comboState) == m3States.end()){
+            m3.add_state(comboState);
+            if(m1.is_in_final_state() || m2.is_in_final_state()) { m3.add_final_state(comboState); }
+            fsm::fill(m1, m2, m3, comboState);
+        }
+        m3.add_transition_rule(prevState, i, comboState);
+        m1 = m1old;
+        m2 = m2old;
+    }
+}
+
+std::ostream &fsm::FSM::ins(std::ostream &out) const {
+    int stateC = get_states_count(), alphaC = get_alphabet_count();
+    auto table = get_transition_table();
+    fsm::State st;
+
+    for(int i = 0; i < stateC; i++){
+        out << states_[i] << " | ";
+        for(int j = 0; j < alphaC; j++){
+            st = table[i][j];
+            out << st << "\t";
+        }
+        out << "\n";
+    }
+
+    return out;
+}
+
+fsm::FSM fsm::FSM::operator|(const fsm::FSM &rhs) const {
+    fsm::FSM unionMachine = *this & rhs;
+    auto thisFinalStates = get_final_states();
+    auto otherFinalStates = rhs.get_final_states();
+    auto unionEndStates = unionMachine.get_final_states();
+    std::vector<fsm::State> newEndStates;
+
+
+    for(int i = 0, sz1 = get_final_states_count(); i < sz1; i++){
```

Comment:
If I'm not mistaken, because you're using `||` on line 236 the `unionMachine` (defined here from the intersection) already has the correct final states for a union machine. Hence the loops are not needed, because `newEndStates` would be the same to `unionEndStates`.

If I'm not mistaken the final states for a union machine should be the the states `<s1, s2>` where **either** `s1` is a final state in `this` **or** `s2` is a final state in `rhs`.
For an intersection machine the final states should be he states `<s1, s2>` where both `s1` and `s2` are final states.

It seems to me that you've swapped the two definitions.

Reply:
Oh no, you are right, I have indeed swapped them. The current "operator&" actually implements the Union between two machines, instead of the intersection. The operand symbols and some names need to be changes. It is a HUGE mistake, Thank you so much for noticing <3, and I am so sorry.